### DISCUSSION TITLE
Redefine vjp only for certain parameters

### DISCUSSION THREAD

**Discussion by alessandrosimon** [google/flax#2722] [Q&A]
Hello. I would like to implement a standard dense layer that is also masked. The masking function is just a sigmoid of logits that enter as parameters into the model.
There are a couple of difficulties that I encountered:

1. Originally I wanted to reuse `nn.Dense` for the linear part. But a as I need to multiply the weight matrix / kernel by the mask and the kernel is not accessible in a (compact) `__call__` function when `nn.Dense` is a class member, I guess the only way to implement the masking is by copying the `nn.Dense` implementation and adding the masking logic into the `__call__`. Is there a cleaner way?
2. As for the main part of the question, after adding the logits as `self.param` I want to override the `vjp` w.r.t to this parameter, while the others should stay the same. I tried defining a new `nn.Module` with my original module as a member and then using `nn.custom_vjp`. I can access the correct parameters in the parameter dictionary. How does one proceed now in the best way? Unfreeze -> modify -> freeze?
3. For the backward pass I also need the value of the function evaluated at the given input. When defining a custom jax.fwd/jax.bwd one can just add the relevant value to the residue list and reuse it for the backward pass. But in this case I'm using the `nn.vjp` on the member module in the fwd function because I'm not quite sure how to write it manually in the case of a `nn.Module`
4. And last but not least, the backward pass also depends on the value of the random sample that was drawn during the forward pass. What would be the best way to access that information?

I'm also attaching the code for the masked dense layer and the vjp wrapper module.
Hope, what I wrote makes sense.

```python
class dense_masked(nn.Module):
    features: int
    drop_rate: float = 0.5

    def init_zphi(self, drop_rate):
        def init_fn(key, shape):
            zero = np.log(1 - drop_rate) - np.log(drop_rate)
            return random.normal(key, shape) + zero

        return init_fn

    @nn.compact
    def __call__(self, inputs, deterministic=False):
        kernel = self.param(
            "kernel", nn.initializers.lecun_normal(), (inputs.shape[-1], self.features)
        )

        zphi = self.param(
            "zphi",
            self.init_zphi(self.drop_rate),
            kernel.shape,
        )

    rng = self.make_rng("l0")
    u = self.variable("l0", "u", nn.zeros, rng, kernel.shape)

    if not deterministic:
        pi = nn.sigmoid(zphi)
        usamp = random.uniform(rng, kernel.shape)
        z = np.where(usamp < pi, 1.0, 0.0)
        u.value = usamp
        kernel = kernel * z
    else:
        pi = nn.sigmoid(zphi)
        z = np.where(pi > 0.0, 1.0, 0.0)
        kernel = kernel * z

    y = lax.dot_general(
            inputs,
            kernel,
            (((inputs.ndim - 1,), (0,)), ((), ())),
    )

    return y


class test(nn.Module):
    @nn.compact
    def __call__(self, x):
        dm = dense_masked(9)
        dm(x)

        def fwd(mdl, x):
            # store info about f(x)?
            return nn.vjp(lambda mdl, x: mdl(x), mdl, x)

        def bwd(vjp_fn, y_t):
            params_t, *inputs_t = vjp_fn(y_t)
            # get info about u that was sampled
            # only change paramter tanget for zphi?
            return (params_t, *inputs_t)

        newvjp = nn.custom_vjp(lambda mdl, x: mdl(x), forward_fn=fwd, backward_fn=bwd)

        return np.sum(newvjp(dm, x))
```

**Comment by cgarciae** [comment]
Hey! Its an interesting question. I'll try to answer the first point for now.

Here a way to mask the parameters of a `Dense` Module using `nn.map_variables`:

```python
import flax.linen as nn
import jax.numpy as jnp
import jax

class MaskedDense(nn.Module):
    @nn.compact
    def __call__(self, x, mask):
        def apply_mask(variables):
            if not self.is_initializing():
                variables['params']['kernel'] *= mask
            return variables

        _MaskedDense = nn.map_variables(
            nn.Dense, mapped_collections="params", 
            trans_in_fn=apply_mask, mutable="params")

        return _MaskedDense(3, name="masked_dense")(x)

x = jnp.ones((1, 2))
mask = jnp.zeros((2, 3)) # mask kernel with zeros, output should be zeros
module = MaskedDense()
variables = module.init(jax.random.PRNGKey(0), x, mask)
y = module.apply(variables, x, mask)

print(variables['params'])
print(y)
```
Output:
```
FrozenDict({
    masked_dense: {
        bias: Array([0., 0., 0.], dtype=float32),
        kernel: Array([[ 0.8859115 ,  0.8155951 ,  0.44417906],
               [-1.4494084 , -1.2475327 , -0.5755586 ]], dtype=float32),
    },
})
[[0. 0. 0.]] # works!
```
I've used a predefined `mask` array for the kernel in this example but you can customize the `apply_mask` function to do what you what (e.g. apply the `sigmoid` function).

**Reply by alessandrosimon** [reply]
Thanks, that's exactly what I was looking for!

**Reply by alessandrosimon** [reply]
Small follow-up question. Is there a way to initialize the variable `mask` from within `__call__` with the correct size?

**Reply by cgarciae** [reply]
I used `mask` being an input as an example, but you can probably just create an array within `apply_mask` directly.

**Comment by alessandrosimon** [comment]
Having read the documentation a bit more I think the rest of the questions are also resolved. I'm not sure if this is the most canonical (in flax) way to do it but it seems to work.

```python3

class wrapper(nn.Module):
    @nn.compact
    def __call__(self, x):
        dm = dense_masked(9)
        dm(x)
        
        def fwd(mdl, x):
            vjpret = nn.vjp(lambda mdl, x: mdl(x), mdl, x)
            # carry holds information needed for the backward pass
            carry = (mdl.variables['l0']['u'], vjpret[0])
            return (vjpret[0], (vjpret[1], carry))

        def bwd(res, y_t):
            # unpack the residuals
            vjp_fn, r = res
            params_t, *inputs_t = vjp_fn(y_t)
            
            # use carry to compute dzphi
            unfrozen = inputs_t[0]['params'].unfreeze()
            # for testing set to one
            unfrozen['zphi'] = np.ones((3,9))
            inputs_t[0]['params'] = freeze(unfrozen)
            
            return (params_t, *inputs_t)

        newvjp = nn.custom_vjp(
            lambda mdl, x: mdl(x), forward_fn=fwd, backward_fn=bwd)
       
        return np.sum(newvjp(dm, x)[0])
```

It seems that the documentation in [flax docs - model surgery](https://flax.readthedocs.io/en/latest/guides/model_surgery.html) is slightly outdated because the proposed way of flattening and unflattening the parameters doesn't work anymore. 
Anyways, I think for my use case it is more straightforward and cleaner to just modify the grad directly in the training loop. That shouldn't break anything right?
